### PR TITLE
APPS-2043 collect images

### PR DESCRIPTION
**Pull request by Gvaihir**
Added module `collect_images.py` which interacts with `nextaur:collect` and the subclasses of `ImageRef` added in  https://github.com/dnanexus/dx-toolkit/pull/1175. 

This PR depends on https://github.com/dnanexus/dx-toolkit/pull/1175

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -12,6 +12,7 @@ Categories for each release: Added, Changed, Deprecated, Removed, Fixed, Securit
 
 * `dx extract_assay expression`
 * Return `jobLogsForwardingStatus` field in `dx describe --verbose job-xxxx`
+* Docker image can be cached on the platform
```
Comment (YuxinShi0423) [inline CHANGELOG.md:15]: 1. Are users able to turn caching off? 
2. What if the docker image needs the applet runner to provide docker credential, but the applet builder does not have such access?

**Review by YuxinShi0423 — CHANGES_REQUESTED** [review]

File: src/python/dxpy/nextflow/ImageRef.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+            filename=file_name,
+            project=config["DX_PROJECT_CONTEXT_ID"],
+            folder=self._caching_dir,
+            name=file_name,
+            parents=True,
+            properties={"image_digest": self._digest}
+        )
+        return uploaded_dx_file.get_id()
+
+    def _reconstruct_image_ref(self):
```
[the remark this answers is not in the document]
Comment (Gvaihir) [reply src/python/dxpy/nextflow/ImageRef.py:131]: See [spec](https://confluence.internal.dnanexus.com/pages/viewpage.action?pageId=113675686)
TL;DR - it should be referenced by `image_name@digest` (see your own reference, it can't be just digest). In this case tarball name won't have the version, but it will have a digest property. See how `DockerImage` in nextaur solves that.

File: CHANGELOG.md (hunk 1 above)
Comment (Gvaihir) [reply CHANGELOG.md:15]: 1. Yes. Storage costs are negligible compared to amount of data produced by each run.
2. That's up to users to manage creds.

File: src/python/dxpy/nextflow/ImageRef.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (YuxinShi0423) [reply src/python/dxpy/nextflow/ImageRef.py:131]:
```
        full_ref = self._join_if_exists(":", [repo_and_image_name, self._tag])
        return full_ref
 ```
 
 I am saying here the _tag could be empty, because instead digest is used to render the full_ref together with repo and image name. Like docker.io/library/ubuntu@sha256:26c68657ccce2cb0a31b330cb0be2b5e108d467f641c62e13ab40cbec258c68d


But you only include the case for docker.io/library/ubuntu:22.04

**Review by YuxinShi0423 — CHANGES_REQUESTED** [review]

File: src/python/dxpy/nextflow/ImageRef.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Gvaihir) [reply src/python/dxpy/nextflow/ImageRef.py:131]: yes, but what is the concern here? Is it not covered with tests or this is not explicit in the docs/spec? Could you be more specific in what your expectation is and what you think needs a change?

[the remark this answers is not in the document]
Comment (YuxinShi0423) [reply src/python/dxpy/nextflow/ImageRef.py:131]: I am worried if the digest format is specified by the user, then we cannot get the full_ref here. 

if the user gives `docker.io/library/ubuntu@sha256:26c68657ccce2cb0a31b330cb0be2b5e108d467f641c62e13ab40cbec258c68d`, then the full_ref generated here is only `docker.io/library/ubuntu`

**Review by YuxinShi0423 — CHANGES_REQUESTED** [review]

File: src/python/dxpy/nextflow/ImageRef.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Gvaihir) [reply src/python/dxpy/nextflow/ImageRef.py:131]: ok, why is it necessary to get a full reference here?

[the remark this answers is not in the document]
Comment (Gvaihir) [reply src/python/dxpy/nextflow/ImageRef.py:131]: fixed [8c73ccd](https://github.com/dnanexus/dx-toolkit/pull/1263/commits/8c73ccd0056d414f4685e6c09b13cefb59639c7d)

**Review by r-i-v-a — APPROVED** [review]

File: src/python/dxpy/nextflow/ImageRef.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Gvaihir) [reply src/python/dxpy/nextflow/ImageRef.py:131]: @YuxinShi0423 please resolve if this is satisfied

[the remark this answers is not in the document]
Comment (YuxinShi0423) [reply src/python/dxpy/nextflow/ImageRef.py:131]: Thanks! that looks good to me.

File: src/python/dxpy/nextflow/ImageRef.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        """
+        Docker image reference has the form of <REPOSITORY_NAME>/<IMAGE_NAME>:<VERSION_TAG>
+        """
```
Comment (YuxinShi0423) [inline src/python/dxpy/nextflow/ImageRef.py:134]: The docstring should be updated as well?

Comment (Gvaihir) [reply src/python/dxpy/nextflow/ImageRef.py]: Sure, fixed in [3a84ddc](https://github.com/dnanexus/dx-toolkit/pull/1263/commits/3a84ddc6932d82affeca9622538a1a7524144a71)

**Comment by Gvaihir** [comment]
@YuxinShi0423 - wherever you requested the changes please approve. Your request for changes blocks the merge.

**Review by YuxinShi0423 — APPROVED** [review]

File: src/python/dxpy/nextflow/ImageRef.py (hunk 3 above)
Comment (YuxinShi0423) [reply src/python/dxpy/nextflow/ImageRef.py]: Thanks!
